### ISSUE TITLE
Mobs, animals and NPCs

### ISSUE THREAD

**Issue by MarkuBu** [D00Med/farlands#165]
Minetest really needs a new system for mobs. There is absolutely no mod that really works well.

Here are some prerequisites I would have:
- Mobs should rotate smooth, not like they do now
- Animals shouldn't spawn and despawn continuously on a random event. They should spawn like in MC on map generation time in groups from 1-4 and stay there until they are killed or die. 
- Traders should also spawn on map generation 
- NPC's shouldn't spawn if a building is generated, one or maybe randomly two per building.
- Animals should spawn based on the biome, not based on ground nodes

This is to prevent the map from being flooded with animals and NPC's.

![screenshot_20171113_102906](https://user-images.githubusercontent.com/16163058/32943117-64ff6758-cb8b-11e7-8361-6937221d800f.png)

It is not a problem to survive on such a map because you always have enough to eat. Just wait for the next animal.

- Hostile mobs should spawn on random events
- They should despawn after a certain time, by time of day, by light level or until they are killed.
 
Will add more later

**Comment by tobyplowy** [comment]
@MarkuBu i wish that was a thin but the mt core devs dount give a crap about mobs so we probably wount see them untill 1.0.0 :/

well makeing a new mob api will be good but its too much work :/

**Comment by MarkuBu** [comment]
The core devs want to have a mob engine, but there is no engine that is nearly good enough to be added. Also it is necessary to implement some functions to the engine. And as long as nobody knows what is really needed they can't add anything. 

mobs redo is the best mob engine we currently have, but from my point of view it is still horrible. Sofars entity_ai has some good ideas, but it is far away from good or even finished.

**Comment by D00Med** [comment]
Things NPC's should do:
>Use doors
>Adopt empty homes.
>Change to different types of NPC (eg: butcher, smith, etc.) depending on what is inside their home.
>Go to their homes at night

Things that would be nice for them to do:
>Use beds
>Stick around the general area of their home. (maybe teleport if they are too far away)

**Comment by NewbProgrammer101** [comment]
@hkzorman 's advanced_npc mod does pretty much all of these things.

**Comment by MarkuBu** [comment]
Well, First of all, NPC's shouldn't spawn on time based events, cause they will "flood" the map. Spawners should only be used for hostile mobs e.g. in dungeons

@toby, can you make different textures for NPC's for different jobs and maybe male/female?

And is it possible to detach the head from the keyframe animation and rotate the head from a script with bone animation? I know that it is possible to move each bone (see ObjectRef in lua_api.txt) but I don't know how a model has to be prepared for this.

**Comment by tobyplowy** [comment]
> Well, First of all, NPC's shouldn't spawn on time based events, cause they will "flood" the map. Spawners should only be used for hostile mobs e.g. in dungeons

I agree

> @toby, can you make different textures for NPC's for different jobs and maybe male/female?

Yup can do

> And is it possible to detach the head from the keyframe animation and rotate the head from a script with bone animation? I know that it is possible to move each bone (see ObjectRef in lua_api.txt) but I don't know how a model has to be prepared for this.

Is this the thing you have to do to rotate the head so it looks at the player if so that would be awesome to have

**Comment by D00Med** [comment]
>Well, First of all, NPC's shouldn't spawn on time based events, cause they will "flood" the map. Spawners should only be used for hostile mobs e.g. in dungeons

I agree also, should the "spawners" instead spawn an NPC when they are right-clicked?

>And is it possible to detach the head from the keyframe animation and rotate the head from a script with bone animation? I know that it is possible to move each bone (see ObjectRef in lua_api.txt) but I don't know how a model has to be prepared for this.

It's possible I think, but I don't know how.

**Comment by MarkuBu** [comment]
> should the "spawners" instead spawn an NPC when they are right-clicked?

Why a spawner at all? Spawn one villager per building. If you need more villager, build a house and breed a new one or find a second villager

**Comment by D00Med** [comment]
@MarkuBu The main purpose of the spawners is to allow players to populate villages. 
By having it work as it did, there was more randomness to the population of villages, but I'll admit it didn't work very well at all.  Have you ever tried getting a mob to go where you want?  breeding NPC's (apart from being a little creepy, breeding them and kidnapping the kids xD), will make it awkward to get the NPC's where you want them. A spawner is a nice easy way to get around this.

**Comment by MarkuBu** [comment]
But "spawn" as many villagers as you want for free is totally overpowered and doesn't add any progress to the game.

**Comment by D00Med** [comment]
Again, this is something that could be avoided with automatic spawning, if not for the fact that it doesn't stop. I don't see it as a major problem anyway, the npc's are still not particularly useful unless they have houses. Maybe players should be able to obtain a tool that spawns a limited number of npc's before it wears out? or maybe spawners, once placed by a player, are unbreakeable until after they spawn a certain number of NPC's, and then are removed.

**Comment by NewbProgrammer101** [comment]
Maybe define the parameters of the spawners by the amount of village plots placed. This is easy to do.

**Comment by D00Med** [comment]
That doesn't sound easy. 
Remember the spawners are mostly for use with player-built villages and towns, not really for the villages placed by the mapgen
